### ISSUE TITLE
Recording a game with Misdirection causes infinite recursion

### ISSUE THREAD

**Issue by anka-213** [danielyule/hearthbreaker#96]
Both `replay.py` and `Misdirection` overloads the `game.current_player.agent.choose_target`, which under some circumstances causes an infinite recursion.

To replicate:
```
diff --git a/tests/card_tests/hunter_tests.py b/tests/card_tests/hunter_tests.py
index e8dc293..006b0db 100644
--- a/tests/card_tests/hunter_tests.py
+++ b/tests/card_tests/hunter_tests.py
@@ -7,6 +7,7 @@ from tests.agents.testing_agents import CardTestingAgent, OneCardPlayingAgent, W
     PlayAndAttackAgent, SelfSpellTestingAgent
 from tests.testing_utils import generate_game_for, mock
 from hearthbreaker.cards import *
+from hearthbreaker.replay import *
 
 
 class TestHunter(unittest.TestCase):
@@ -271,6 +272,7 @@ class TestHunter(unittest.TestCase):
 
     def test_Misdirection(self):
         game = generate_game_for(Misdirection, StonetuskBoar, CardTestingAgent, PlayAndAttackAgent)
+        record(game)
 
         for turn in range(0, 4):
             game.play_single_turn()
```

Traceback: (cut off after first full cycle)
```
Traceback (most recent call last):
  File "/media/shared/projects/hearthbreaker/hearthbreaker/tests/card_tests/hunter_tests.py", line 278, in test_Misdirection
    game.play_single_turn()
  File "/media/shared/projects/hearthbreaker/hearthbreaker/hearthbreaker/engine.py", line 141, in play_single_turn
    self.current_player.agent.do_turn(self.current_player)
  File "/media/shared/projects/hearthbreaker/hearthbreaker/tests/agents/testing_agents.py", line 102, in do_turn
    minion.attack()
  File "/media/shared/projects/hearthbreaker/hearthbreaker/hearthbreaker/game_objects.py", line 894, in attack
    super().attack()
  File "/media/shared/projects/hearthbreaker/hearthbreaker/hearthbreaker/game_objects.py", line 406, in attack
    target = self.choose_target(targets)
  File "/media/shared/projects/hearthbreaker/hearthbreaker/hearthbreaker/game_objects.py", line 434, in choose_target
    return self.player.choose_target(targets)
  File "/media/shared/projects/hearthbreaker/hearthbreaker/hearthbreaker/engine.py", line 447, in choose_target
    return self.agent.choose_target(targets)
  File "/media/shared/projects/hearthbreaker/hearthbreaker/hearthbreaker/replay.py", line 386, in choose_target
    target = self.agent.choose_target(targets)
  File "/media/shared/projects/hearthbreaker/hearthbreaker/hearthbreaker/cards/spells/hunter.py", line 136, in choose_random
    old_target = old_target_func(targets)
  File "/media/shared/projects/hearthbreaker/hearthbreaker/hearthbreaker/replay.py", line 386, in choose_target
    target = self.agent.choose_target(targets)
  File "/media/shared/projects/hearthbreaker/hearthbreaker/hearthbreaker/cards/spells/hunter.py", line 136, in choose_random
    old_target = old_target_func(targets)
  File "/media/shared/projects/hearthbreaker/hearthbreaker/hearthbreaker/replay.py", line 386, in choose_target
    target = self.agent.choose_target(targets)
  etc...
```

**Comment by danielyule** [comment]
I think this should be fixed with 33c2d0397462b4fe12a6bb003b14316044ed85c4.  Feel free to test it and let me know if there's still a problem.

**Comment by anka-213** [comment]
Yes, it definitely fixed it.
